File: LeetCode/Largest Rectangle In Histogram/Largest Rectangle In Histogram.cpp
#include <bits/stdc++.h> //you can skip this when writting on leetcode

using namespace std; //you can skip this when writting on leetcode

class Solution {
public:
    int largestRectangleArea(vector<int>& a) {
        int i,n=a.size(),maxi=0,height,width;
        stack<int> st;
        for(i=0;i<=n;i++){
            while(st.size() && ((i==n) || (a[st.top()]>=a[i]))){
                height=a[st.top()]; st.pop();
                if(st.size()==0) width=i;
                else width=i-st.top()-1;
                maxi=max(maxi, (height*width));
            }
            st.push(i);
        }
        return maxi;
    }
};

File: LeetCode/Largest Rectangle In Histogram/Largest Rectangle In Histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        ans = 0
        l = [(-1, 0)] # here the list l will serve as the stack 

        for i, j in enumerate(heights + [0]):
            if l[-1][1] < j:
                l.append((i, j))
            else:
                while l[-1][1] > j:
                    pre_i,pre_j = l.pop()
                    ans = max(ans, pre_j * (i - pre_i))
                if l[-1][1] < j:
                    l.append((pre_i, j))
        return ans
